jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java

```diff
@@ -82,12 +84,63 @@ protected void doStop() throws Exception
     @Override
     public CompletableFuture<Void> preCreateConnections(int connectionCount)
     {
-        CompletableFuture<?>[] futures = new CompletableFuture[connectionCount];
-        for (int i = 0; i < connectionCount; i++)
+        if (LOG.isDebugEnabled())
+            LOG.debug("Precreating connections {}/{}", connectionCount, getMaxConnectionCount());
+
+        List<CompletableFuture<?>> futures = new ArrayList<>();
+        loop : for (int i = 0; i < connectionCount; i++)
         {
-            futures[i] = tryCreateAsync(getMaxConnectionCount());
+            Pool<Connection>.Entry entry = pool.reserve();
+            while (entry == null)
+            {
+                if (pool.size() >= pool.getMaxEntries())
+                    break loop;
+                if (pool.getMaxMultiplex() <= 1)
+                    throw new IllegalStateException();
+                entry = pool.reserve();
+            }
+
+            final Pool<Connection>.Entry reserved = entry;
+
+            Promise.Completable<Connection> future = new Promise.Completable<Connection>()
```

Comment:
From here the code is duplicated somewhat... but I won't factor out until we are agreed on basic approach.
The tryCreate method was inlined because of the logic needed above for multiplexed connections